File: app/src/main/java/com/sysu/edu/preference/EditPreference.java
package com.sysu.edu.preference;

import static com.sysu.edu.api.CommonUtil.trim;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;

import com.sysu.edu.R;
import com.sysu.edu.databinding.PreferenceEditBinding;

public class EditPreference extends Preference {
    
    private String mValue;
    
    public EditPreference(@NonNull Context context) {
        this(context, null);
    }
    
    public EditPreference(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, android.R.attr.editTextPreferenceStyle);
    }
    
    public EditPreference(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }
    
    public EditPreference(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        setLayoutResource(R.layout.preference_edit);
    }
    
    @Override
    public void onBindViewHolder(@NonNull PreferenceViewHolder holder) {
        PreferenceEditBinding binding = PreferenceEditBinding.bind(holder.itemView);
        binding.textInputLayout.setHint(getTitle());
        binding.textInputLayout.setStartIconDrawable(getIcon());
        binding.textField.setText(getValue());
        binding.textField.addTextChangedListener(new TextWatcher() {
            @Override
            public void afterTextChanged(Editable s) {
            
            }
            
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            
            }
            
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                setValue(s.toString());
            }
        });
    }
    
    public String getValue() {
        return mValue;
    }
    
    public void setValue(String value) {
        mValue = value;
        persistString(value);
    }
    
    @Override
    protected void onSetInitialValue(@Nullable Object defaultValue) {
        setValue(getPersistedString(trim((String) defaultValue)));
    }
}

File: app/src/main/java/com/sysu/edu/preference/EditPreference.kt
package com.sysu.edu.preference

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.sysu.edu.R
import com.sysu.edu.api.CommonUtil.trim
import com.sysu.edu.databinding.PreferenceEditBinding

class EditPreference(context: Context,
                     attrs: AttributeSet? = null,
                     defStyleAttr: Int = android.R.attr.editTextPreferenceStyle,
                     defStyleRes: Int = 0) : Preference(context, attrs, defStyleAttr, defStyleRes) {
	constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0, 0)
	
	private var mValue: String? = null
	
	init {
		layoutResource = R.layout.preference_edit
	}
	
	override fun onBindViewHolder(holder: PreferenceViewHolder) {
		PreferenceEditBinding.bind(holder.itemView).apply {
			textInputLayout.hint = title
			textInputLayout.startIconDrawable = getIcon()
			textField.setText(value)
			textField.addTextChangedListener(object : TextWatcher {
				override fun afterTextChanged(s: Editable?) {
				}
				
				override fun beforeTextChanged(s: CharSequence?,
				                               start: Int,
				                               count: Int,
				                               after: Int) {
				}
				
				override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
					value = "$s"
				}
			})
		}
	}
	
	var value: String?
		get() = mValue
		set(value) {
			mValue = value
			persistString(value)
		}
	
	override fun onSetInitialValue(defaultValue: Any?) {
		value = getPersistedString(trim(defaultValue as String?))
	}
}
